host/src/request.rs

```diff
@@ -1,77 +1,331 @@
+    #[structopt(long, require_equals = true)]
+    pub bonsai: bool,
+    #[structopt(long, require_equals = true)]
+    pub snark: bool,
+    #[structopt(long, require_equals = true)]
+    pub profile: bool,
+    #[structopt(long, require_equals = true)]
+    pub execution_po2: i32,
 }
 
-#[derive(Debug, Serialize)]
-pub struct JsonRpcResponseError {
-    pub jsonrpc: String,
-    pub id: serde_json::Value,
-    pub error: JsonRpcError,
+impl FromStr for Risc0Params {
+    type Err = HostError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        serde_json::from_str(s).map_err(|e| e.into())
+    }
 }
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct JsonRpcResponse<T> {
-    pub jsonrpc: String,
-    pub id: serde_json::Value,
-    pub result: Option<T>,
+#[derive(StructOpt, Default, Clone, Serialize, Deserialize, Debug, ToSchema)]
+#[serde(default)]
+/// A partial proof request config.
+pub struct ProofRequestOpt {
+    #[structopt(long, require_equals = true)]
+    /// The block number for the block to generate a proof for.
+    pub block_number: Option<u64>,
+    #[structopt(long, require_equals = true)]
+    /// RPC URL for retreiving block by block number.
+    pub rpc: Option<String>,
+    #[structopt(long, require_equals = true)]
+    /// The L1 node URL for signal root verify and get txlist info from proposed
+    /// transaction.
+    pub l1_rpc: Option<String>,
+    #[structopt(long, require_equals = true)]
+    /// The beacon node URL for retreiving data blobs.
+    pub beacon_rpc: Option<String>,
+    #[structopt(long, require_equals = true)]
+    /// The network to generate the proof for.
+    pub network: Option<String>,
+    #[structopt(long, require_equals = true)]
+    // Graffiti.
+    pub graffiti: Option<String>,
+    #[structopt(long, require_equals = true)]
+    /// The protocol instance data.
+    pub prover: Option<String>,
+    #[structopt(long, require_equals = true)]
+    /// The proof type.
+    pub proof_type: Option<String>,
+    #[structopt(long, require_equals = true)]
+    /// The sgx prover params in JSON format.
+    pub sgx: Option<SgxParams>,
+    #[structopt(long, require_equals = true)]
+    /// The risc0 prover params in JSON format.
+    pub risc0: Option<Risc0Params>,
 }
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct JsonRpcRequest<T: Serialize> {
-    pub jsonrpc: String,
-    pub id: serde_json::Value,
-    pub method: String,
-    pub params: T,
+impl ProofRequestOpt {
+    /// Read a partial proof request config from a file.
+    pub fn from_file<T>(path: T) -> Result<Self, HostError>
+    where
+        T: AsRef<Path>,
+    {
+        let file = std::fs::File::open(path)?;
+        let reader = std::io::BufReader::new(file);
+        let config: serde_json::Value = serde_json::from_reader(reader)?;
+        Self::deserialize(&config).map_err(|e| e.into())
+    }
+
+    /// Merge a partial proof request into current one.
+    pub fn merge(&mut self, other: &Self) {
```

Comment:
Any reason you manually merge each value together here? Compared to previously by merging directly on the json file so each value is automatically merged without any manual work. I understand it's maybe a bit unconventional, but each value is automatically supported this way which I think is pretty nice, can never forget to add a value here for any value added in the future.

It also looks like it's not possible to override the SGX/risc0 settings anymore using the json rpc request?

Reply:
We can do the custom merge that was here before, I was just avoiding doing serialization and deserialization multiple times on the same data. 

I'll add SGX and risc0 settings support for proof requests, although I want to first do a nicer implementation for those since it doesn't seem practical to keep adding custom config fields for each new prover, rather just have a generic `special_args` or `prover_args` field that accepts a `Option<Value>` and pass it to the prover. This would however require a slight change in the prover itself, since it would now be expecting those extra params inside of the new field rather than sgx/risc0 fields.

@Brechtpd Any thoughts on this idea?

Reply:
https://github.com/taikoxyz/raiko/pull/31/commits/2ac321e0c0f72b8b2a470812922e522f6cd718a3 Addresses the idea I just outlined

Reply:
Ah this gets it closer to how it worked before I believe. One reason I still used sgx/risc0 names instead of something like a general `prover_args` name is so that the user can still have the default values stored in the config file for each prover separately. So with only `prover_args` I think it's not possible for a user to set the default values for for example both SGX and SP1 in the config file (or even specify them in the json rpc for when raiko perhaps in the future supports generating proofs together for muliple proof types).

And yeah saw that you moved some prover specific data into main host app, which honestly was okay to me. But now that it's opaque again to the host, I think it does require the json style merge (which another nice property of it, works even for data that the host doesn't know!) because currently it is kind of limited because either all prover params are overwritten or none of them are, which is a bit strange I think because it works differently from the other values.

Reply:
This is changed by this function https://github.com/taikoxyz/raiko/blob/9e558d7083ba60d7bd49d15487ffdfdece0ed2eb/host/src/request.rs#L207

Reply:
Prover args are now just a flattened struct that can support simultaneous multi prover configs and pass them as expected to the underlying provers